File: backend/app/providers/places/overpass.py
"""Nearby places from OpenStreetMap via the Overpass API.

Enabled with ``PLACES_PROVIDER=overpass``. Kept off by default: the public
Overpass endpoint rate-limits aggressively and would make report submission
depend on a third party. On any failure this falls back to the seeded
catalogue, so scoring still completes.
"""

from __future__ import annotations

import logging

import httpx
from sqlalchemy.orm import Session

from app.core.enums import PlaceType
from app.providers.places.base import NearbyPlaceResult
from app.providers.places.seeded import SeededPlacesProvider
from app.services.geo import haversine_meters

logger = logging.getLogger("roadwatch.places")

#: OSM tag -> our facility taxonomy.
_TAG_MAP: list[tuple[str, str, PlaceType]] = [
    ("amenity", "hospital", PlaceType.HOSPITAL),
    ("amenity", "clinic", PlaceType.HOSPITAL),
    ("amenity", "school", PlaceType.SCHOOL),
    ("amenity", "college", PlaceType.SCHOOL),
    ("amenity", "fire_station", PlaceType.EMERGENCY_SERVICE),
    ("amenity", "police", PlaceType.EMERGENCY_SERVICE),
    ("highway", "bus_stop", PlaceType.BUS_STOP),
]


class OverpassPlacesProvider:
    name = "overpass"

    def __init__(self, db: Session, api_url: str, timeout_seconds: float = 8.0) -> None:
        self.db = db
        self.api_url = api_url
        self.timeout_seconds = timeout_seconds
        self._fallback = SeededPlacesProvider(db)

    def find_nearby(
        self, latitude: float, longitude: float, radius_meters: int
    ) -> list[NearbyPlaceResult]:
        try:
            payload = self._query(latitude, longitude, radius_meters)
        except (httpx.HTTPError, ValueError) as exc:
            logger.warning("Overpass unavailable (%s); using the seeded catalogue", exc)
            return self._fallback.find_nearby(latitude, longitude, radius_meters)

        results: list[NearbyPlaceResult] = []
        for element in payload.get("elements", []):
            tags = element.get("tags", {})
            place_type = _classify(tags)
            if place_type is None:
                continue
            lat = element.get("lat") or element.get("center", {}).get("lat")
            lon = element.get("lon") or element.get("center", {}).get("lon")
            if lat is None or lon is None:
                continue
            distance = haversine_meters(latitude, longitude, float(lat), float(lon))
            if distance > radius_meters:
                continue
            results.append(
                NearbyPlaceResult(
                    place_type=place_type,
                    name=tags.get("name") or place_type.value.replace("_", " ").title(),
                    latitude=float(lat),
                    longitude=float(lon),
                    distance_meters=round(distance, 1),
                    source=self.name,
                )
            )

        results.sort(key=lambda item: item.distance_meters)
        return results

    def _query(self, latitude: float, longitude: float, radius_meters: int) -> dict:
        clauses = "".join(
            f'node(around:{radius_meters},{latitude},{longitude})["{key}"="{value}"];'
            for key, value, _ in _TAG_MAP
        )
        query = f"[out:json][timeout:{int(self.timeout_seconds)}];({clauses});out center;"
        with httpx.Client(timeout=self.timeout_seconds) as client:
            response = client.post(self.api_url, data={"data": query})
            response.raise_for_status()
            return response.json()


def _classify(tags: dict) -> PlaceType | None:
    for key, value, place_type in _TAG_MAP:
        if tags.get(key) == value:
            return place_type
    return None

File: supabase/functions/api/providers/places/overpass.ts
// Ported from backend/app/providers/places/overpass.py. Off by default (the
// public Overpass endpoint rate-limits aggressively); falls back to the
// seeded catalogue on any failure.

import type { SupabaseClient } from "@supabase/supabase-js";
import type { NearbyPlaceResult, NearbyPlacesProvider } from "./base.ts";
import { haversineMeters } from "../../services/geo.ts";
import type { PlaceType } from "../../_shared/enums.ts";
import { SeededPlacesProvider } from "./seeded.ts";

const TAG_MAP: [string, string, PlaceType][] = [
  ["amenity", "hospital", "HOSPITAL"],
  ["amenity", "clinic", "HOSPITAL"],
  ["amenity", "school", "SCHOOL"],
  ["amenity", "college", "SCHOOL"],
  ["amenity", "fire_station", "EMERGENCY_SERVICE"],
  ["amenity", "police", "EMERGENCY_SERVICE"],
  ["highway", "bus_stop", "BUS_STOP"],
];

function classify(tags: Record<string, string>): PlaceType | null {
  for (const [key, value, placeType] of TAG_MAP) {
    if (tags[key] === value) return placeType;
  }
  return null;
}

export class OverpassPlacesProvider implements NearbyPlacesProvider {
  name = "overpass";
  private fallback: SeededPlacesProvider;

  constructor(client: SupabaseClient, private apiUrl: string, private timeoutMs = 8_000) {
    this.fallback = new SeededPlacesProvider(client);
  }

  async findNearby(latitude: number, longitude: number, radiusMeters: number): Promise<NearbyPlaceResult[]> {
    let payload: { elements?: Record<string, unknown>[] };
    try {
      payload = await this.query(latitude, longitude, radiusMeters);
    } catch {
      return await this.fallback.findNearby(latitude, longitude, radiusMeters);
    }

    const results: NearbyPlaceResult[] = [];
    for (const element of payload.elements ?? []) {
      const tags = (element.tags as Record<string, string>) ?? {};
      const placeType = classify(tags);
      if (!placeType) continue;

      const center = element.center as Record<string, unknown> | undefined;
      const lat = (element.lat as number | undefined) ?? (center?.lat as number | undefined);
      const lon = (element.lon as number | undefined) ?? (center?.lon as number | undefined);
      if (lat === undefined || lon === undefined) continue;

      const distance = haversineMeters(latitude, longitude, lat, lon);
      if (distance > radiusMeters) continue;

      results.push({
        placeType,
        name: (tags.name as string) || placeType.replace(/_/g, " ").replace(/\b\w/g, (c) => c.toUpperCase()),
        latitude: lat,
        longitude: lon,
        distanceMeters: Math.round(distance * 10) / 10,
        source: this.name,
      });
    }

    results.sort((a, b) => a.distanceMeters - b.distanceMeters);
    return results;
  }

  private async query(latitude: number, longitude: number, radiusMeters: number): Promise<{ elements?: Record<string, unknown>[] }> {
    const clauses = TAG_MAP.map(
      ([key, value]) => `node(around:${radiusMeters},${latitude},${longitude})["${key}"="${value}"];`,
    ).join("");
    const timeoutSeconds = Math.floor(this.timeoutMs / 1000);
    const query = `[out:json][timeout:${timeoutSeconds}];(${clauses});out center;`;

    const response = await fetch(this.apiUrl, {
      method: "POST",
      body: new URLSearchParams({ data: query }),
      signal: AbortSignal.timeout(this.timeoutMs),
    });
    if (!response.ok) throw new Error(`Overpass HTTP ${response.status}`);
    return await response.json();
  }
}
